File: library/src/main/java/com/raizlabs/android/dbflow/sql/trigger/CompletedTrigger.java
package com.raizlabs.android.dbflow.sql.trigger;

import com.raizlabs.android.dbflow.config.BaseDatabaseDefinition;
import com.raizlabs.android.dbflow.config.FlowManager;
import com.raizlabs.android.dbflow.sql.Query;
import com.raizlabs.android.dbflow.sql.QueryBuilder;
import com.raizlabs.android.dbflow.sql.SqlUtils;
import com.raizlabs.android.dbflow.structure.Model;

/**
 * Description: The last piece of a TRIGGER statement, this class contains the BEGIN...END and the logic in between.
 */
public class CompletedTrigger<ModelClass extends Model> implements Query {

    /**
     * The first pieces of this TRIGGER statement
     */
    private TriggerMethod<ModelClass> mTriggerMethod;

    /**
     * The query to run between the BEGIN and END of this statement
     */
    private Query mTriggerLogicQuery;

    CompletedTrigger(TriggerMethod<ModelClass> triggerMethod, Query triggerLogicQuery) {
        mTriggerMethod = triggerMethod;
        mTriggerLogicQuery = triggerLogicQuery;
    }

    @Override
    public String getQuery() {
        QueryBuilder queryBuilder = new QueryBuilder(mTriggerMethod.getQuery());
        queryBuilder.append("\nBEGIN")
                .append("\n").append(mTriggerLogicQuery.getQuery()).append(";")
                .append("\nEND");
        return queryBuilder.getQuery();
    }


    /**
     * Turns on this trigger
     */
    public void enable() {
        BaseDatabaseDefinition databaseDefinition = FlowManager.getDatabaseForTable(mTriggerMethod.mOnTable);
        databaseDefinition.getWritableDatabase().execSQL(getQuery());
    }

    /**
     * Disables this trigger
     */
    public void disable() {
        SqlUtils.dropTrigger(mTriggerMethod.mOnTable, mTriggerMethod.mTrigger.mTriggerName);
    }
}

File: dbflow-common/src/main/kotlin/com/raizlabs/dbflow5/query/CompletedTrigger.kt
package com.raizlabs.dbflow5.query

import com.raizlabs.dbflow5.database.DatabaseWrapper
import com.raizlabs.dbflow5.dropTrigger
import com.raizlabs.dbflow5.sql.Query

/**
 * Description: The last piece of a TRIGGER statement, this class contains the BEGIN...END and the logic in between.
 */
class CompletedTrigger<TModel: Any> internal constructor(
    /**
     * The first pieces of this TRIGGER statement
     */
    private val triggerMethod: TriggerMethod<TModel>, triggerLogicQuery: Query) : Query {

    /**
     * The query to run between the BEGIN and END of this statement
     */
    private val triggerLogicQuery = arrayListOf<Query>()

    override val query: String
        get() =
            "${triggerMethod.query}\nBEGIN\n${triggerLogicQuery.joinToString(separator = ";\n")};\nEND"

    init {
        this.triggerLogicQuery.add(triggerLogicQuery)
    }

    /**
     * Appends the nextStatement to this query as another line to be executed by trigger.
     */
    infix fun and(nextStatement: Query) = apply {
        this.triggerLogicQuery.add(nextStatement)
    }


    /**
     * Turns on this trigger
     */
    fun enable(databaseWrapper: DatabaseWrapper) {
        databaseWrapper.execSQL(query)
    }

    /**
     * Disables this trigger
     */
    fun disable(databaseWrapper: DatabaseWrapper) {
        dropTrigger(databaseWrapper, triggerMethod.trigger.name)
    }
}
